### PR TITLE
AP_InertialSensor: add driver for imu sensor SCHA63T

### PR DESCRIPTION
**Pull request by jfbblue0922**
This is the IMU (SCHA63T) driver for JFB110（the new flight controller of JAE）.
This driver was tested on JFB100 hardware.

### CODE REVIEW

File: libraries/AP_InertialSensor/AP_InertialSensor_SCHA63T.h [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+  gyro, which means it needs two SPIDevice pointers
+ */
+#pragma once
+
+#include <AP_HAL/AP_HAL.h>
+
+#include "AP_InertialSensor.h"
+#include "AP_InertialSensor_Backend.h"
+
+static constexpr int16_t combine(uint8_t msb, uint8_t lsb)
```
Comment (tridge) [inline libraries/AP_InertialSensor/AP_InertialSensor_SCHA63T.h:26]: please put this in the cpp, so it isn't exposed to other cpp files

File: libraries/AP_InertialSensor/AP_InertialSensor_SCHA63T.cpp [hunk 2]
```diff
@@ -0,0 +1,451 @@
...
+    WITH_SEMAPHORE(dev_due->get_semaphore());
+
+    // error initialise is OK
+    ret_scha63t = true;
+
+    // wait 25ms for non-volatile memory (NVM) read
+    hal.scheduler->delay(25);
+
+    // set DUE operation mode on (must be less than 1ms)
+    ret_scha63t &= RegisterWrite(SCHA63T_DUE, MODE, MODE_NORM);
```
Comment (tridge) [inline libraries/AP_InertialSensor/AP_InertialSensor_SCHA63T.cpp:123]: why call this twice?

File: libraries/AP_InertialSensor/AP_InertialSensor_SCHA63T.cpp [hunk 3]
```diff
@@ -0,0 +1,451 @@
...
+    ret_scha63t &= RegisterWrite(SCHA63T_UNO, G_FILT_DYN, G_FILT);
+    ret_scha63t &= RegisterWrite(SCHA63T_UNO, A_FILT_DYN, A_FILT);
+
+    // reset DUE write (0001h) to register 18h
+    ret_scha63t &= RegisterWrite(SCHA63T_DUE, RESCTRL, HW_RES);
+    // wait 25ms for non-volatile memory (NVM) read
+    hal.scheduler->delay(25);
+
+    // set DUE operation mode on (must be less than 1ms)
+    ret_scha63t &= RegisterWrite(SCHA63T_DUE, MODE, MODE_NORM);
```
Comment (tridge) [inline libraries/AP_InertialSensor/AP_InertialSensor_SCHA63T.cpp:140]: again why twice? if its needed then add a comment as to why

File: libraries/AP_InertialSensor/AP_InertialSensor_SCHA63T.cpp [hunk 4]
```diff
@@ -0,0 +1,451 @@
...
+        // system in FAILURE mode (startup_attempt not equl 0 startup_attempt = 1)
+        // reset UNO write (0001h) to register 18h
+        ret_scha63t &= RegisterWrite(SCHA63T_UNO, RESCTRL, HW_RES);
+        // reset DUE write (0001h) to register 18h
+        ret_scha63t &= RegisterWrite(SCHA63T_DUE, RESCTRL, HW_RES);
+        // wait 25ms for non-volatile memory (NVM) read
+        hal.scheduler->delay(25);
+
+        // set DUE operation mode on (must be less than 1ms)
+        ret_scha63t &= RegisterWrite(SCHA63T_DUE, MODE, MODE_NORM);
```
Comment (tridge) [inline libraries/AP_InertialSensor/AP_InertialSensor_SCHA63T.cpp:159]: another double call

File: libraries/AP_InertialSensor/AP_InertialSensor_SCHA63T.cpp [hunk 5]
```diff
@@ -0,0 +1,451 @@
...
+    uint8_t rsp_accl_z[3];
+    uint8_t rsp_temper[3];
+
+    int16_t accel_x = 0;
+    int16_t accel_y = 0;
+    int16_t accel_z = 0;
+    int16_t uno_temp = 0;
+
+    // ACCL_X Cmd Send (This Response rsp_accl_x is Dust!!)
+    ret_scha63t &= RegisterRead(SCHA63T_UNO, ACC_X, rsp_accl_x);
```
Comment (tridge) [inline libraries/AP_InertialSensor/AP_InertialSensor_SCHA63T.cpp:244]: first command is lost? Is that what you mean by "is Dust" ?

File: libraries/AP_InertialSensor/AP_InertialSensor_SCHA63T.cpp [hunk 6]
```diff
@@ -0,0 +1,451 @@
...
+    // ACCL_Y Cmd Send + ACCL_X Response Receive
```
Comment (tridge) [inline libraries/AP_InertialSensor/AP_InertialSensor_SCHA63T.cpp:245]: can this be done as a single read block of all 3 registers?

**Review by tridge — CHANGES_REQUESTED** [review]
we should squash into a single AP_InertialSensor commit - if you are not familiar with how to do that with git I can do it for you

File: libraries/AP_InertialSensor/AP_InertialSensor_SCHA63T.cpp [hunk 7]
```diff
@@ -0,0 +1,451 @@
...
+        ret = dev_due->transfer(buf, 4, buf, 4);
+        memcpy(val, buf, 4);
+        break;
+    default:
+        break;
+    }
+
+    if (ret == true) {
+        bCrc = crc8_sae(val, 3);
+        if ( bCrc != val[3] ) {
```
Comment (tridge) [inline libraries/AP_InertialSensor/AP_InertialSensor_SCHA63T.cpp:412]: this accesses beyond the end of the val[] array, as callers declare the array as size 3 on the stack

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_InertialSensor/AP_InertialSensor_SCHA63T.h (hunk 1 above)
Comment (jfbblue0922) [reply libraries/AP_InertialSensor/AP_InertialSensor_SCHA63T.h]: Thank you.
I changed it.

File: libraries/AP_InertialSensor/AP_InertialSensor_SCHA63T.cpp (hunk 2 above)
Comment (jfbblue0922) [reply libraries/AP_InertialSensor/AP_InertialSensor_SCHA63T.cpp:128]: The sensor datasheet says to operate twice in the power up sequence.
figure1

File: libraries/AP_InertialSensor/AP_InertialSensor_SCHA63T.cpp (hunk 6 above)
Comment (jfbblue0922) [reply libraries/AP_InertialSensor/AP_InertialSensor_SCHA63T.cpp:250]: Yes.
you are right.

File: libraries/AP_InertialSensor/AP_InertialSensor_SCHA63T.cpp (hunk 7 above)
Comment (jfbblue0922) [reply libraries/AP_InertialSensor/AP_InertialSensor_SCHA63T.cpp:417]: Thank you.
I changed it.

**Review by jfbblue0922 — COMMENTED** [review]
All file checked.

**Comment by jfbblue0922** [comment]
> we should squash into a single AP_InertialSensor commit - if you are not familiar with how to do that with git I can do it for you

We don't know about git very well, so please do it for us.
Thank you for your cooperation.

**Review by jfbblue0922 — COMMENTED** [review]
We finished to change code.
Please review them and let us know what should we do next.

**Comment by rmackay9** [comment]
@jfbblue0922 are you OK with us closing this now?

**Pull request closed without merging**